File: backtest/src/pdt-tracker.ts
/**
 * Pattern Day Trader (PDT) tracker for backtest simulation.
 * Mirrors OCaml pdt_tracker.ml logic: max 3 day trades in rolling 5 business days
 * when equity is below $25,000.
 */

export interface DayTradeRecord {
  symbol: string;
  date: number; // timestamp ms
}

export interface PdtTracker {
  trades: DayTradeRecord[];
  equity: number;
}

const FIVE_BUSINESS_DAYS_MS = 5 * 24 * 60 * 60 * 1000;
const PDT_EQUITY_THRESHOLD = 25_000;
const MAX_DAY_TRADES_RESTRICTED = 3;

/**
 * Create an empty PDT tracker with the given equity.
 */
export function createPdtTracker(equity: number): PdtTracker {
  return { trades: [], equity };
}

/**
 * Remove trades older than 5 business days from the current time.
 */
function pruneOldTrades(currentTime: number, trades: DayTradeRecord[]): DayTradeRecord[] {
  return trades.filter(t => (currentTime - t.date) < FIVE_BUSINESS_DAYS_MS);
}

/**
 * Record a day trade (same-day open and close of a position).
 */
export function addDayTrade(tracker: PdtTracker, symbol: string, date: number): PdtTracker {
  const trade: DayTradeRecord = { symbol, date };
  const trades = pruneOldTrades(date, [trade, ...tracker.trades]);
  return { ...tracker, trades };
}

/**
 * Update the equity value used for PDT determination.
 */
export function updateEquity(tracker: PdtTracker, equity: number): PdtTracker {
  return { ...tracker, equity };
}

/**
 * Check if we can make another day trade.
 * Unlimited if equity >= $25K, otherwise max 3 in rolling 5-day window.
 */
export function canDayTrade(tracker: PdtTracker): boolean {
  return tradesRemaining(tracker) > 0;
}

/**
 * Number of day trades remaining in the current window.
 * Returns 999 if unrestricted (equity >= $25K).
 */
export function tradesRemaining(tracker: PdtTracker): number {
  if (tracker.equity >= PDT_EQUITY_THRESHOLD) return 999;
  const used = tracker.trades.length;
  return Math.max(0, MAX_DAY_TRADES_RESTRICTED - used);
}

/**
 * Current count of day trades in the rolling window.
 */
export function dayTradeCount(tracker: PdtTracker): number {
  return tracker.trades.length;
}

/**
 * PDT status matching OCaml variant type.
 */
export type PdtStatus =
  | { kind: 'unrestricted' }
  | { kind: 'restricted'; tradesUsed: number }
  | { kind: 'blocked' };

export function status(tracker: PdtTracker): PdtStatus {
  if (tracker.equity >= PDT_EQUITY_THRESHOLD) {
    return { kind: 'unrestricted' };
  }
  const used = tracker.trades.length;
  if (used >= MAX_DAY_TRADES_RESTRICTED) {
    return { kind: 'blocked' };
  }
  return { kind: 'restricted', tradesUsed: used };
}

/**
 * Convert to OCaml-compatible pdt_tracker record for passing to risk.evaluate.
 * The OCaml module expects { trades: linked-list, equity: float }.
 */
export function toOCaml(tracker: PdtTracker): { trades: unknown; equity: number } {
  // Build a Melange-style linked list from the trades array
  let list: unknown = 0; // /* [] */ 0
  for (let i = tracker.trades.length - 1; i >= 0; i--) {
    list = {
      hd: { symbol: tracker.trades[i].symbol, date: tracker.trades[i].date },
      tl: list,
    };
  }
  return { trades: list, equity: tracker.equity };
}

File: trading-core/lib/pdt_tracker.ml
(** Pattern Day Trader tracking.
    Tracks day trades in a rolling 5-business-day window.
    PDT rule: 4+ day trades in 5 business days with equity < $25K = restricted. *)

type day_trade = {
  symbol: Symbol.t;
  date: float;  (** Unix timestamp ms *)
  strategy: Types.strategy_id;
}

type t = {
  trades: day_trade list;
  equity: float;
}

let empty equity = { trades = []; equity }

(** 5 business days in milliseconds (approximate: 5 * 24 * 60 * 60 * 1000) *)
let five_business_days_ms = 5.0 *. 24.0 *. 60.0 *. 60.0 *. 1000.0

let prune_old_trades (current_time : float) (trades : day_trade list) : day_trade list =
  List.filter (fun (trade : day_trade) ->
    let age = current_time -. trade.date in
    age < five_business_days_ms
  ) trades

let add_trade (tracker : t) (trade : day_trade) : t =
  let trades = prune_old_trades trade.date (trade :: tracker.trades) in
  { tracker with trades }

let day_trade_count (tracker : t) : int =
  List.length tracker.trades

let trades_remaining (tracker : t) : int =
  if tracker.equity >= 25000.0 then
    999  (* effectively unlimited *)
  else
    let used = day_trade_count tracker in
    Int.max 0 (3 - used)

let can_day_trade (tracker : t) : bool =
  trades_remaining tracker > 0

let status (tracker : t) : Types.pdt_status =
  if tracker.equity >= 25000.0 then Types.Unrestricted
  else
    let used = day_trade_count tracker in
    if used >= 3 then Types.Blocked
    else Types.Restricted { trades_used = used }
